# **FRAMEWORK V7 DATASET MAESTRO - ÍNDICE DE PRIORIZACIÓN DE VARIABLES PARA MACHINE LEARNING (IPML)**

## **M0. Configuración**

In [1]:
import pandas as pd

print("="*90)
print("SCRIPT")
print("ÍNDICE DE PERTINENCIA PARA MACHINE LEARNING (IPML)")
print("="*90)
print()

url_universo_analitico_vif = "https://raw.githubusercontent.com/jriatiga/FRAMEWORK_V7/refs/heads/main/DATA/MASTER/C10_MASTER/Resultados_EDA/Analisis_Multicolinealidad/universo_analitico_vif.csv"

universo = pd.read_csv(
    url_universo_analitico_vif
)


SCRIPT
ÍNDICE DE PERTINENCIA PARA MACHINE LEARNING (IPML)



## **M1. Universo Analitico**

In [2]:
print()

print("="*90)
print("UNIVERSO ANALÍTICO")
print("="*90)

print()

print(f"Variables cargadas : {len(universo)}")

display(universo.head())


UNIVERSO ANALÍTICO

Variables cargadas : 35


,Variable,Tipo,Relaciones_Fuertes,Indice_Redundancia,Nivel_Redundancia,Diagnostico_Redundancia,VIF,Estado_VIF,Diagnostico_VIF,Recomendacion_VIF
0,Precipitacion_mm,Numérica,0,0.000,Muy Baja,Variable prácticamente independiente.,1.898,Baja,Variable con baja dependencia estructural.,Variable estable para Machine Learning.
1,Temp_Max_C,Numérica,1,0.167,Muy Baja,Variable prácticamente independiente.,31.915,Muy Alta,Comparte gran parte de su información con otra...,Revisar antes del entrenamiento de modelos.
2,Temp_Min_C,Numérica,0,0.000,Muy Baja,Variable prácticamente independiente.,10.573,Alta,Presenta una dependencia considerable con otra...,Analizar conjuntamente con otras variables.
3,Temp_Media_C,Numérica,1,0.167,Muy Baja,Variable prácticamente independiente.,36.048,Muy Alta,Comparte gran parte de su información con otra...,Revisar antes del entrenamiento de modelos.
4,Humedad_Relativa,Numérica,0,0.000,Muy Baja,Variable prácticamente independiente.,7.128,Moderada,Existe una dependencia moderada aceptable para...,Sin restricciones importantes.


¿Por qué solo este archivo?

Porque ya contiene:

* Tipo
* Redundancia
* Diagnóstico
* VIF
* Estado VIF
* Recomendaciones

Es decir, ya reúne el conocimiento construido hasta ahora.

In [3]:
#==========================================================================================
# VALIDACIÓN DEL UNIVERSO ANALÍTICO
#==========================================================================================

columnas = [

    "Variable",

    "Indice_Redundancia",

    "Estado_VIF"

]

faltantes = [

    c

    for c in columnas

    if c not in universo.columns

]

if len(faltantes)>0:

    raise Exception(

        f"Faltan columnas: {faltantes}"

    )

print()

print("="*90)
print("VALIDACIÓN")
print("="*90)

print()

print("Universo Analítico validado correctamente.")


VALIDACIÓN

Universo Analítico validado correctamente.


In [4]:
#==========================================================================================
# PUNTAJE DE REDUNDANCIA
#==========================================================================================

universo["Puntaje_Redundancia"] = (

    1 -

    universo["Indice_Redundancia"]

).round(3)

print()

print("="*90)
print("PUNTAJE DE REDUNDANCIA")
print("="*90)

display(

    universo[

        [

            "Variable",

            "Indice_Redundancia",

            "Puntaje_Redundancia"

        ]

    ].head()

)


PUNTAJE DE REDUNDANCIA


,Variable,Indice_Redundancia,Puntaje_Redundancia
0,Precipitacion_mm,0.000,1.000
1,Temp_Max_C,0.167,0.833
2,Temp_Min_C,0.000,1.000
3,Temp_Media_C,0.167,0.833
4,Humedad_Relativa,0.000,1.000


In [5]:
#==========================================================================================
# ESCALA DEL PUNTAJE DE MULTICOLINEALIDAD
#==========================================================================================

escala_vif = {

    "Baja": 1.00,

    "Moderada": 0.75,

    "Alta": 0.50,

    "Muy Alta": 0.25,

    "Perfecta": 0.00,

    "Sin Multicolinealidad": 1.00,

    "No Calculado": 0.50

}

Justificación metodológica

Debido a que el Factor de Inflación de la Varianza (VIF) presenta una escala no lineal y valores potencialmente infinitos, se empleó la clasificación cualitativa obtenida para construir un puntaje normalizado entre 0 y 1. Esta transformación facilita su integración con las demás dimensiones del IPML.


In [6]:
#==========================================================================================
# PUNTAJE DE MULTICOLINEALIDAD
#==========================================================================================

universo["Puntaje_Multicolinealidad"] = (

    universo["Estado_VIF"]

    .map(escala_vif)

).round(3)

print()

print("="*90)
print("PUNTAJE DE MULTICOLINEALIDAD")
print("="*90)

display(

    universo[

        [

            "Variable",

            "Estado_VIF",

            "Puntaje_Multicolinealidad"

        ]

    ].head(10)

)


PUNTAJE DE MULTICOLINEALIDAD


,Variable,Estado_VIF,Puntaje_Multicolinealidad
0,Precipitacion_mm,Baja,1.00
1,Temp_Max_C,Muy Alta,0.25
2,Temp_Min_C,Alta,0.50
3,Temp_Media_C,Muy Alta,0.25
4,Humedad_Relativa,Moderada,0.75
5,Velocidad_Viento,Baja,1.00
6,Radiacion_Solar,Baja,1.00
7,Anio,Moderada,0.75
8,Mes,Baja,1.00
9,ONI,Baja,1.00


Observación

**Puntajes de Redundancia y Multicolinealidad por Variable**

| Variable | IR | Puntaje R | Estado VIF | Puntaje M |
| :--- | :--- | :--- | :--- | :--- |
| **Precipitación** | 0.000 | 1.000 | Baja | 1.000 |
| **Temp_Max** | 0.167 | 0.833 | Muy Alta | 0.250 |

Eso significa que:

* Temp_Max no será descartada automáticamente.

* Simplemente tendrá un puntaje menor en una dimensión.

Eso es exactamente lo que queríamos.

* El IPML no elimina variables.

* Las prioriza.



In [7]:

#==========================================================================================
# CARGA DEL CATÁLOGO DE CONOCIMIENTO
#==========================================================================================

url_catalogo_conocimiento = "https://raw.githubusercontent.com/jriatiga/FRAMEWORK_V7/refs/heads/main/DATA/MASTER/C10_MASTER/Preparacion_Machine_Learning/catalogo_conocimiento/catalogo_conocimiento_variables.csv"


catalogo = pd.read_csv(

    url_catalogo_conocimiento,
    sep = ';',
    encoding = 'latin1',
    engine = 'python'

)

print()

print("="*90)
print("CATÁLOGO DE CONOCIMIENTO")
print("="*90)

print()

print(f"Variables catalogadas : {len(catalogo)}")

display(catalogo.head())


CATÁLOGO DE CONOCIMIENTO

Variables catalogadas : 35


,Variable,Categoria,Importancia,Peso,Justificacion
0,Precipitacion_mm,Climática,Muy Alta,3,Principal variable que controla la disponibili...
1,Temp_Max_C,Climática,Muy Alta,3,Influye en los procesos de evaporación y balan...
2,Temp_Min_C,Climática,Muy Alta,3,Describe la variabilidad térmica diaria y sus ...
3,Temp_Media_C,Climática,Muy Alta,3,Representa el comportamiento térmico general d...
4,Humedad_Relativa,Climática,Muy Alta,3,Relacionada con evapotranspiración y formación...


In [8]:
#==========================================================================================
# VALIDACIÓN DEL CATÁLOGO
#==========================================================================================

columnas = [

    "Variable",

    "Peso"

]

faltantes = [

    c

    for c in columnas

    if c not in catalogo.columns

]

if len(faltantes)>0:

    raise Exception(

        f"Faltan columnas: {faltantes}"

    )

print()

print("="*90)
print("VALIDACIÓN")
print("="*90)

print()

print("Catálogo validado correctamente.")


VALIDACIÓN

Catálogo validado correctamente.


In [9]:
#==========================================================================================
# VALIDACIÓN DEL CATÁLOGO
#==========================================================================================

url = "https://raw.githubusercontent.com/jriatiga/FRAMEWORK_V7/refs/heads/main/DATA/MASTER/C10_MASTER/Resultados_EDA/Analisis_Multicolinealidad/universo_analitico_vif.csv"

universo = pd.read_csv(url)

# Verificar la carga
universo.head()

,Variable,Tipo,Relaciones_Fuertes,Indice_Redundancia,Nivel_Redundancia,Diagnostico_Redundancia,VIF,Estado_VIF,Diagnostico_VIF,Recomendacion_VIF
0,Precipitacion_mm,Numérica,0,0.000,Muy Baja,Variable prácticamente independiente.,1.898,Baja,Variable con baja dependencia estructural.,Variable estable para Machine Learning.
1,Temp_Max_C,Numérica,1,0.167,Muy Baja,Variable prácticamente independiente.,31.915,Muy Alta,Comparte gran parte de su información con otra...,Revisar antes del entrenamiento de modelos.
2,Temp_Min_C,Numérica,0,0.000,Muy Baja,Variable prácticamente independiente.,10.573,Alta,Presenta una dependencia considerable con otra...,Analizar conjuntamente con otras variables.
3,Temp_Media_C,Numérica,1,0.167,Muy Baja,Variable prácticamente independiente.,36.048,Muy Alta,Comparte gran parte de su información con otra...,Revisar antes del entrenamiento de modelos.
4,Humedad_Relativa,Numérica,0,0.000,Muy Baja,Variable prácticamente independiente.,7.128,Moderada,Existe una dependencia moderada aceptable para...,Sin restricciones importantes.


In [10]:
#==========================================================================================
# INTEGRACIÓN DEL CONOCIMIENTO DEL DOMINIO
#==========================================================================================

universo = universo.merge(

    catalogo[
        [
            "Variable",
            "Categoria",
            "Importancia",
            "Peso"
        ]
    ],

    on="Variable",

    how="left"

)

print()

print("="*90)
print("INTEGRACIÓN DEL CATÁLOGO")
print("="*90)

print()

print(f"Variables del Universo : {len(universo)}")

print()

print("Valores faltantes después de la integración:")

print(

    universo[
        [
            "Categoria",
            "Importancia",
            "Peso"
        ]

    ].isna().sum()

)

display(universo.head())


INTEGRACIÓN DEL CATÁLOGO

Variables del Universo : 35

Valores faltantes después de la integración:
Categoria      0
Importancia    0
Peso           0
dtype: int64


,Variable,Tipo,Relaciones_Fuertes,Indice_Redundancia,Nivel_Redundancia,Diagnostico_Redundancia,VIF,Estado_VIF,Diagnostico_VIF,Recomendacion_VIF,Categoria,Importancia,Peso
0,Precipitacion_mm,Numérica,0,0.000,Muy Baja,Variable prácticamente independiente.,1.898,Baja,Variable con baja dependencia estructural.,Variable estable para Machine Learning.,Climática,Muy Alta,3
1,Temp_Max_C,Numérica,1,0.167,Muy Baja,Variable prácticamente independiente.,31.915,Muy Alta,Comparte gran parte de su información con otra...,Revisar antes del entrenamiento de modelos.,Climática,Muy Alta,3
2,Temp_Min_C,Numérica,0,0.000,Muy Baja,Variable prácticamente independiente.,10.573,Alta,Presenta una dependencia considerable con otra...,Analizar conjuntamente con otras variables.,Climática,Muy Alta,3
3,Temp_Media_C,Numérica,1,0.167,Muy Baja,Variable prácticamente independiente.,36.048,Muy Alta,Comparte gran parte de su información con otra...,Revisar antes del entrenamiento de modelos.,Climática,Muy Alta,3
4,Humedad_Relativa,Numérica,0,0.000,Muy Baja,Variable prácticamente independiente.,7.128,Moderada,Existe una dependencia moderada aceptable para...,Sin restricciones importantes.,Climática,Muy Alta,3


In [11]:
#==========================================================================================
# PUNTAJE DE CONOCIMIENTO DEL DOMINIO
#==========================================================================================

peso_maximo = catalogo["Peso"].max()

universo["Puntaje_Conocimiento_Dominio"] = (

    universo["Peso"] /

    peso_maximo

).round(3)

print()

print("="*90)
print("PUNTAJE DE CONOCIMIENTO DEL DOMINIO")
print("="*90)

display(

    universo[
        [
            "Variable",
            "Importancia",
            "Peso",
            "Puntaje_Conocimiento_Dominio"
        ]
    ].head(10)

)


PUNTAJE DE CONOCIMIENTO DEL DOMINIO


,Variable,Importancia,Peso,Puntaje_Conocimiento_Dominio
0,Precipitacion_mm,Muy Alta,3,1.000
1,Temp_Max_C,Muy Alta,3,1.000
2,Temp_Min_C,Muy Alta,3,1.000
3,Temp_Media_C,Muy Alta,3,1.000
4,Humedad_Relativa,Muy Alta,3,1.000
5,Velocidad_Viento,Alta,2,0.667
6,Radiacion_Solar,Muy Alta,3,1.000
7,Anio,Media,1,0.333
8,Mes,Alta,2,0.667
9,ONI,Muy Alta,3,1.000


In [12]:
#==========================================================================================
# CARGA DEL ANÁLISIS DE CALIDAD
#==========================================================================================

url_calidad = "https://raw.githubusercontent.com/jriatiga/FRAMEWORK_V7/refs/heads/main/DATA/MASTER/C10_MASTER/Calidad_Variables_Dataset_Maestro_Framework_v04.csv"

calidad = pd.read_csv(url_calidad)

print()

print("="*90)
print("ANÁLISIS DE CALIDAD")
print("="*90)

print()

print(f"Variables evaluadas : {len(calidad)}")

display(calidad.head())


ANÁLISIS DE CALIDAD

Variables evaluadas : 42


,Variable,Tipo,Registros,Completos,Nulos,% Cobertura,% Nulos,Valores Únicos,Duplicados,% Ceros,Calificación,Recomendación
0,Nodo,object,528,528,0,100.0,0.0,4,524,NaN,A - Excelente,Lista para modelado
1,Fecha,datetime64[ns],528,528,0,100.0,0.0,132,396,NaN,A - Excelente,Lista para modelado
2,Precipitacion_mm,float64,528,528,0,100.0,0.0,437,91,0.19,A - Excelente,Lista para modelado
3,Temp_Max_C,float64,528,528,0,100.0,0.0,406,122,0.00,A - Excelente,Lista para modelado
4,Temp_Min_C,float64,528,528,0,100.0,0.0,410,118,0.00,A - Excelente,Lista para modelado


In [13]:
#==========================================================================================
# INTEGRACIÓN DEL ANÁLISIS DE CALIDAD
#==========================================================================================

universo = universo.merge(

    calidad[
        [
            "Variable",
            "% Cobertura"
        ]
    ],

    on="Variable",

    how="left"

)

print()

print("="*90)
print("INTEGRACIÓN DE CALIDAD")
print("="*90)

print()

print(

    universo[
        ["% Cobertura"]
    ].isna().sum()

)

display(universo.head())


INTEGRACIÓN DE CALIDAD

% Cobertura    0
dtype: int64


,Variable,Tipo,Relaciones_Fuertes,Indice_Redundancia,Nivel_Redundancia,Diagnostico_Redundancia,VIF,Estado_VIF,Diagnostico_VIF,Recomendacion_VIF,Categoria,Importancia,Peso,Puntaje_Conocimiento_Dominio,% Cobertura
0,Precipitacion_mm,Numérica,0,0.000,Muy Baja,Variable prácticamente independiente.,1.898,Baja,Variable con baja dependencia estructural.,Variable estable para Machine Learning.,Climática,Muy Alta,3,1.0,100.0
1,Temp_Max_C,Numérica,1,0.167,Muy Baja,Variable prácticamente independiente.,31.915,Muy Alta,Comparte gran parte de su información con otra...,Revisar antes del entrenamiento de modelos.,Climática,Muy Alta,3,1.0,100.0
2,Temp_Min_C,Numérica,0,0.000,Muy Baja,Variable prácticamente independiente.,10.573,Alta,Presenta una dependencia considerable con otra...,Analizar conjuntamente con otras variables.,Climática,Muy Alta,3,1.0,100.0
3,Temp_Media_C,Numérica,1,0.167,Muy Baja,Variable prácticamente independiente.,36.048,Muy Alta,Comparte gran parte de su información con otra...,Revisar antes del entrenamiento de modelos.,Climática,Muy Alta,3,1.0,100.0
4,Humedad_Relativa,Numérica,0,0.000,Muy Baja,Variable prácticamente independiente.,7.128,Moderada,Existe una dependencia moderada aceptable para...,Sin restricciones importantes.,Climática,Muy Alta,3,1.0,100.0


In [14]:
#==========================================================================================
# PUNTAJE DE CALIDAD
#==========================================================================================

universo["Puntaje_Calidad"] = (

    universo["% Cobertura"] /

    100

).round(3)

print()

print("="*90)
print("PUNTAJE DE CALIDAD")
print("="*90)

display(

    universo[
        [
            "Variable",
            "% Cobertura",
            "Puntaje_Calidad"
        ]
    ].head(10)

)


PUNTAJE DE CALIDAD


,Variable,% Cobertura,Puntaje_Calidad
0,Precipitacion_mm,100.0,1.0
1,Temp_Max_C,100.0,1.0
2,Temp_Min_C,100.0,1.0
3,Temp_Media_C,100.0,1.0
4,Humedad_Relativa,100.0,1.0
5,Velocidad_Viento,100.0,1.0
6,Radiacion_Solar,100.0,1.0
7,Anio,100.0,1.0
8,Mes,100.0,1.0
9,ONI,100.0,1.0


Estado actual

El Universo Analítico ya contiene:

| Dimensión                | Columna                        | Estado |
| ------------------------ | ------------------------------ | ------ |
| Redundancia              | `Puntaje_Redundancia`          | ✅      |
| Multicolinealidad        | `Puntaje_Multicolinealidad`    | ✅      |
| Conocimiento del Dominio | `Puntaje_Conocimiento_Dominio` | ✅      |
| Calidad                  | `Puntaje_Calidad`              | ✅      |


---

Observación

* El resultado:

  * Variable	Puntaje_Calidad
  * Todas	1.000

  nos dice algo muy interesante.

  * La dimensión Calidad no discrimina entre variables.

  * No es un error.

  Simplemente significa que todas las variables tienen cobertura del 100 %, por lo que, para este conjunto de datos, la calidad no influye en la priorización.



> El análisis de calidad evidenció una cobertura del 100 % para todas las variables incluidas en el Universo Analítico. En consecuencia, el Puntaje de Calidad asignó el valor máximo (1.0) a todas las variables. Aunque esta dimensión no generó diferenciación entre variables en este estudio, se conserva dentro del IPML porque permite que el Framework sea aplicable a otros conjuntos de datos donde existan diferencias de completitud o calidad.


---

**Construcción del IPML**

Una vez consolidadas las cuatro dimensiones en el Universo Analítico, se define la ecuación oficial del **Índice de Priorización para Machine Learning (IPML)**.

---

**Ecuación Definitiva**

$$IPML = 0.30 \times PCD + 0.30 \times PC + 0.20 \times PR + 0.20 \times PM$$

**Donde:**
* **$PCD$**: Puntaje de Conocimiento del Dominio
* **$PC$**: Puntaje de Calidad
* **$PR$**: Puntaje de Redundancia
* **$PM$**: Puntaje de Multicolinealidad

---

**Justificación Técnica de la Ponderación**

Los pesos asignados responden a la estructura y principios del Framework sustentado en la investigación:

| Dimensión | Peso | Justificación |
| :--- | :---: | :--- |
| **Conocimiento del Dominio ($PCD$)** | **30%** | Es la dimensión transversal más importante; representa el criterio experto sobre la relevancia temática y técnica de la variable para la gestión hídrica. |
| **Calidad de Datos ($PC$)** | **30%** | Evalúa la integridad de la variable. Una variable relevante pierde su valor práctico e introduce sesgo si sus datos no son fiables o están incompletos. |
| **Redundancia ($PR$)** | **20%** | Minimiza la duplicidad de información directa entre pares de variables, optimizando la dimensionalidad del conjunto de datos. |
| **Multicolinealidad ($PM$)** | **20%** | Favorece la estabilidad matemática de los modelos finales reduciendo dependencias lineales complejas entre el grupo de variables. |

In [15]:
print(type(universo))
print(universo.shape)

<class 'pandas.core.frame.DataFrame'>
(35, 16)


In [16]:
universo.to_csv(
    "universo_analitico_ipml.csv",
    index=False,
    encoding="utf-8-sig"
)

In [17]:
#==========================================================================================
# EXPORTACIÓN DEL UNIVERSO ANALÍTICO CONSOLIDADO
#==========================================================================================

universo.to_csv(

    "universo_analitico_ipml.csv",

    index=False,

    encoding="utf-8-sig"

)

universo.to_excel(

    "universo_analitico_ipml.xlsx",

    index=False

)

print()

print("="*90)
print("UNIVERSO ANALÍTICO CONSOLIDADO")
print("="*90)

print()

print("Archivos generados:")

print()

print("universo_analitico_ipml.csv")

print("universo_analitico_ipml.xlsx")


UNIVERSO ANALÍTICO CONSOLIDADO

Archivos generados:

universo_analitico_ipml.csv
universo_analitico_ipml.xlsx


## **M3. Calculo Índice de Priorización para Machine Learning (IPML)**

Integrar las cuatro dimensiones obtenidas en los módulos anteriores para calcular el Índice de Priorización para Machine Learning (IPML) y generar el ranking final de variables.

In [18]:
#==========================================================================================
# SCRIPT
# CÁLCULO DEL ÍNDICE DE PRIORIZACIÓN PARA MACHINE LEARNING (IPML)
#==========================================================================================
#
# Objetivo:
#
# Integrar las dimensiones obtenidas durante la preparación del Universo
# Analítico para calcular el Índice de Priorización para Machine Learning (IPML),
# el cual servirá como criterio objetivo para seleccionar las variables que
# conformarán el Dataset definitivo para el entrenamiento de modelos predictivos.
#
# Dimensiones consideradas:
#
# • Redundancia Estructural
# • Multicolinealidad
# • Conocimiento del Dominio
# • Calidad de los Datos
#
# Producto esperado:
#
# • Ranking IPML
# • Variables priorizadas
#
#==========================================================================================

In [19]:
#==========================================================================================
# LIBRERÍAS
#==========================================================================================

import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)

In [21]:
import os
import pandas as pd

# Definir las rutas
url_universo = "https://raw.githubusercontent.com/jriatiga/FRAMEWORK_V7/refs/heads/main/DATA/MASTER/C11_MASTER/Universo_Analitico/universo_analitico_ipml.csv"
ruta_local = "universo_analitico_ipml.csv"  # O la ruta relativa/absoluta donde lo tengas guardado

try:
    # Intentar cargar desde GitHub
    print("Intentando cargar el universo analítico desde GitHub...")
    universo = pd.read_csv(url_universo)
    print("¡Carga exitosa desde GitHub!")
except Exception as e:
    print(f"No se pudo cargar desde GitHub ({e}). Intentando cargar desde el repositorio local...")
    # Cargar desde la ruta local si falla la web
    universo = pd.read_csv(ruta_local)
    print("¡Carga exitosa desde el archivo local!")

print()
print("=" * 90)
print("UNIVERSO ANALÍTICO CONSOLIDADO")
print("=" * 90)
print()

print(f"Variables : {len(universo)}")
print(f"Columnas  : {universo.shape[1]}")

display(universo.head())

Intentando cargar el universo analítico desde GitHub...
No se pudo cargar desde GitHub (HTTP Error 404: Not Found). Intentando cargar desde el repositorio local...
¡Carga exitosa desde el archivo local!

UNIVERSO ANALÍTICO CONSOLIDADO

Variables : 35
Columnas  : 16


,Variable,Tipo,Relaciones_Fuertes,Indice_Redundancia,Nivel_Redundancia,Diagnostico_Redundancia,VIF,Estado_VIF,Diagnostico_VIF,Recomendacion_VIF,Categoria,Importancia,Peso,Puntaje_Conocimiento_Dominio,% Cobertura,Puntaje_Calidad
0,Precipitacion_mm,Numérica,0,0.000,Muy Baja,Variable prácticamente independiente.,1.898,Baja,Variable con baja dependencia estructural.,Variable estable para Machine Learning.,Climática,Muy Alta,3,1.0,100.0,1.0
1,Temp_Max_C,Numérica,1,0.167,Muy Baja,Variable prácticamente independiente.,31.915,Muy Alta,Comparte gran parte de su información con otra...,Revisar antes del entrenamiento de modelos.,Climática,Muy Alta,3,1.0,100.0,1.0
2,Temp_Min_C,Numérica,0,0.000,Muy Baja,Variable prácticamente independiente.,10.573,Alta,Presenta una dependencia considerable con otra...,Analizar conjuntamente con otras variables.,Climática,Muy Alta,3,1.0,100.0,1.0
3,Temp_Media_C,Numérica,1,0.167,Muy Baja,Variable prácticamente independiente.,36.048,Muy Alta,Comparte gran parte de su información con otra...,Revisar antes del entrenamiento de modelos.,Climática,Muy Alta,3,1.0,100.0,1.0
4,Humedad_Relativa,Numérica,0,0.000,Muy Baja,Variable prácticamente independiente.,7.128,Moderada,Existe una dependencia moderada aceptable para...,Sin restricciones importantes.,Climática,Muy Alta,3,1.0,100.0,1.0


In [22]:
print(universo.columns.tolist())

['Variable', 'Tipo', 'Relaciones_Fuertes', 'Indice_Redundancia', 'Nivel_Redundancia', 'Diagnostico_Redundancia', 'VIF', 'Estado_VIF', 'Diagnostico_VIF', 'Recomendacion_VIF', 'Categoria', 'Importancia', 'Peso', 'Puntaje_Conocimiento_Dominio', '% Cobertura', 'Puntaje_Calidad']


In [23]:
#==========================================================================================
# RECONSTRUCCIÓN DEL PUNTAJE DE REDUNDANCIA
#==========================================================================================

universo["Puntaje_Redundancia"] = (

    1 -

    universo["Indice_Redundancia"]

).round(3)

display(

    universo[
        [

            "Variable",

            "Indice_Redundancia",

            "Puntaje_Redundancia"

        ]

    ].head()

)

,Variable,Indice_Redundancia,Puntaje_Redundancia
0,Precipitacion_mm,0.000,1.000
1,Temp_Max_C,0.167,0.833
2,Temp_Min_C,0.000,1.000
3,Temp_Media_C,0.167,0.833
4,Humedad_Relativa,0.000,1.000


In [24]:
#==========================================================================================
# RECONSTRUCCIÓN DEL PUNTAJE DE MULTICOLINEALIDAD
#==========================================================================================

mapa_vif = {

    "Perfecta":0.00,

    "Muy Alta":0.25,

    "Alta":0.50,

    "Moderada":0.75,

    "Baja":1.00,

    "Sin Multicolinealidad":1.00,

    "No Calculado":1.00

}

universo["Puntaje_Multicolinealidad"] = (

    universo["Estado_VIF"]

    .map(mapa_vif)

)

display(

    universo[
        [

            "Variable",

            "Estado_VIF",

            "Puntaje_Multicolinealidad"

        ]

    ].head()

)

,Variable,Estado_VIF,Puntaje_Multicolinealidad
0,Precipitacion_mm,Baja,1.00
1,Temp_Max_C,Muy Alta,0.25
2,Temp_Min_C,Alta,0.50
3,Temp_Media_C,Muy Alta,0.25
4,Humedad_Relativa,Moderada,0.75


In [25]:
print(universo.columns.tolist())

['Variable', 'Tipo', 'Relaciones_Fuertes', 'Indice_Redundancia', 'Nivel_Redundancia', 'Diagnostico_Redundancia', 'VIF', 'Estado_VIF', 'Diagnostico_VIF', 'Recomendacion_VIF', 'Categoria', 'Importancia', 'Peso', 'Puntaje_Conocimiento_Dominio', '% Cobertura', 'Puntaje_Calidad', 'Puntaje_Redundancia', 'Puntaje_Multicolinealidad']


In [26]:
#==========================================================================================
# VALIDACIÓN
#==========================================================================================

columnas_requeridas = [

    "Puntaje_Redundancia",

    "Puntaje_Multicolinealidad",

    "Puntaje_Conocimiento_Dominio",

    "Puntaje_Calidad"

]

faltantes = [

    c

    for c in columnas_requeridas

    if c not in universo.columns

]

if len(faltantes) > 0:

    raise Exception(

        f"Faltan columnas: {faltantes}"

    )

print()

print("="*90)
print("VALIDACIÓN")
print("="*90)

print()

print("Las cuatro dimensiones del IPML fueron encontradas correctamente.")


VALIDACIÓN

Las cuatro dimensiones del IPML fueron encontradas correctamente.


In [27]:
universo.to_csv(

    "universo_analitico_ipml_integracion_dimensiones.csv",

    index=False,

    encoding="utf-8-sig"

)

universo.to_excel(

    "universo_analitico_ipml_integracion_dimensiones.xlsx",

    index=False

)

In [28]:
#==========================================================================================
# PESOS DEL IPML
#==========================================================================================

peso_dominio = 0.40

peso_redundancia = 0.30

peso_multicolinealidad = 0.20

peso_calidad = 0.10

print()

print("="*90)
print("PESOS DEL IPML")
print("="*90)

print()

print(f"Conocimiento del Dominio : {peso_dominio:.2f}")
print(f"Redundancia              : {peso_redundancia:.2f}")
print(f"Multicolinealidad        : {peso_multicolinealidad:.2f}")
print(f"Calidad                  : {peso_calidad:.2f}")

print()

print(f"Suma de pesos            : {peso_dominio + peso_redundancia + peso_multicolinealidad + peso_calidad:.2f}")


PESOS DEL IPML

Conocimiento del Dominio : 0.40
Redundancia              : 0.30
Multicolinealidad        : 0.20
Calidad                  : 0.10

Suma de pesos            : 1.00


In [29]:
#==========================================================================================
# CÁLCULO DEL ÍNDICE DE PRIORIZACIÓN PARA MACHINE LEARNING (IPML)
#==========================================================================================

universo["IPML"] = (

      peso_dominio *

      universo["Puntaje_Conocimiento_Dominio"]

    +

      peso_calidad *

      universo["Puntaje_Calidad"]

    +

      peso_redundancia *

      universo["Puntaje_Redundancia"]

    +

      peso_multicolinealidad *

      universo["Puntaje_Multicolinealidad"]

).round(4)

print()

print("="*90)
print("ÍNDICE DE PRIORIZACIÓN PARA MACHINE LEARNING")
print("="*90)

display(

    universo[

        [

            "Variable",

            "IPML"

        ]

    ].head(10)

)


ÍNDICE DE PRIORIZACIÓN PARA MACHINE LEARNING


,Variable,IPML
0,Precipitacion_mm,1.0000
1,Temp_Max_C,0.7999
2,Temp_Min_C,0.9000
3,Temp_Media_C,0.7999
4,Humedad_Relativa,0.9500
5,Velocidad_Viento,0.8668
6,Radiacion_Solar,1.0000
7,Anio,0.6832
8,Mes,0.8668
9,ONI,1.0000


In [30]:
#==========================================================================================
# ESTADÍSTICAS DEL IPML
#==========================================================================================

print()

print("="*90)
print("ESTADÍSTICAS DEL IPML")
print("="*90)

display(

    universo["IPML"]

    .describe()

)


ESTADÍSTICAS DEL IPML


,IPML
count,34.000000
mean,0.721988
std,0.205226
min,0.366800
25%,0.550350
50%,0.799900
75%,0.867775
max,1.000000


In [31]:
#==========================================================================================
# DISTRIBUCIÓN DEL IPML
#==========================================================================================

print()

print("="*90)
print("DISTRIBUCIÓN DEL IPML")
print("="*90)

display(

    universo["IPML"]

    .value_counts()

    .sort_index()

)


DISTRIBUCIÓN DEL IPML


,count
IPML,
0.3668,5
0.5000,2
0.5260,1
0.5264,1
0.6222,1
0.6260,1
0.6350,1
0.6832,1
0.7231,1


In [32]:
#==========================================================================================
# RANKING DEL IPML
#==========================================================================================

ranking_ipml = (

    universo

    .sort_values(

        "IPML",

        ascending=False

    )

    .reset_index(drop=True)

)

ranking_ipml.insert(

    0,

    "Ranking",

    ranking_ipml.index + 1

)

print()

print("="*90)
print("RANKING DEL IPML")
print("="*90)

display(

    ranking_ipml.head(15)

)


RANKING DEL IPML


,Ranking,Variable,Tipo,Relaciones_Fuertes,Indice_Redundancia,Nivel_Redundancia,Diagnostico_Redundancia,VIF,Estado_VIF,Diagnostico_VIF,Recomendacion_VIF,Categoria,Importancia,Peso,Puntaje_Conocimiento_Dominio,% Cobertura,Puntaje_Calidad,Puntaje_Redundancia,Puntaje_Multicolinealidad,IPML
0,1,Precipitacion_mm,Numérica,0,0.000,Muy Baja,Variable prácticamente independiente.,1.898,Baja,Variable con baja dependencia estructural.,Variable estable para Machine Learning.,Climática,Muy Alta,3,1.000,100.00,1.000,1.000,1.00,1.0000
1,2,ONI,Numérica,0,0.000,Muy Baja,Variable prácticamente independiente.,1.265,Baja,Variable con baja dependencia estructural.,Variable estable para Machine Learning.,Variabilidad Climática,Muy Alta,3,1.000,100.00,1.000,1.000,1.00,1.0000
2,3,Radiacion_Solar,Numérica,0,0.000,Muy Baja,Variable prácticamente independiente.,1.704,Baja,Variable con baja dependencia estructural.,Variable estable para Machine Learning.,Climática,Muy Alta,3,1.000,100.00,1.000,1.000,1.00,1.0000
3,4,Humedad_Relativa,Numérica,0,0.000,Muy Baja,Variable prácticamente independiente.,7.128,Moderada,Existe una dependencia moderada aceptable para...,Sin restricciones importantes.,Climática,Muy Alta,3,1.000,100.00,1.000,1.000,0.75,0.9500
4,5,Nivel_Minimo,Numérica,0,0.000,Muy Baja,Variable prácticamente independiente.,1.027,Baja,Variable con baja dependencia estructural.,Variable estable para Machine Learning.,Hidrológica,Muy Alta,3,1.000,16.48,0.165,1.000,1.00,0.9165
5,6,pH,Numérica,0,0.000,Muy Baja,Variable prácticamente independiente.,2.815,Baja,Variable con baja dependencia estructural.,Variable estable para Machine Learning.,Calidad del Agua,Muy Alta,3,1.000,9.47,0.095,1.000,1.00,0.9095
6,7,DEMANDA QUIMICA DE OXIGENO (DQO),Numérica,0,0.000,Muy Baja,Variable prácticamente independiente.,2.970,Baja,Variable con baja dependencia estructural.,Variable estable para Machine Learning.,Calidad del Agua,Muy Alta,3,1.000,9.28,0.093,1.000,1.00,0.9093
7,8,Temp_Min_C,Numérica,0,0.000,Muy Baja,Variable prácticamente independiente.,10.573,Alta,Presenta una dependencia considerable con otra...,Analizar conjuntamente con otras variables.,Climática,Muy Alta,3,1.000,100.00,1.000,1.000,0.50,0.9000
8,9,AGUAS RESIDUALES TRATADAS,Numérica,1,0.167,Muy Baja,Variable prácticamente independiente.,1.630,Baja,Variable con baja dependencia estructural.,Variable estable para Machine Learning.,Infraestructura,Muy Alta,3,1.000,18.18,0.182,0.833,1.00,0.8681
9,10,Mes,Numérica,0,0.000,Muy Baja,Variable prácticamente independiente.,1.152,Baja,Variable con baja dependencia estructural.,Variable estable para Machine Learning.,Temporal,Alta,2,0.667,100.00,1.000,1.000,1.00,0.8668


In [33]:
#==========================================================================================
# CLASIFICACIÓN DEL IPML
#==========================================================================================

def clasificar_ipml(valor):

    if valor >= 0.90:
        return "Muy Alta"

    elif valor >= 0.80:
        return "Alta"

    elif valor >= 0.70:
        return "Media"

    else:
        return "Baja"


universo["Nivel_IPML"] = universo["IPML"].apply(clasificar_ipml)

print()

print("="*90)
print("CLASIFICACIÓN DEL IPML")
print("="*90)

display(

    universo[
        [
            "Variable",
            "IPML",
            "Nivel_IPML"
        ]
    ].sort_values(
        "IPML",
        ascending=False
    ).head(15)

)


CLASIFICACIÓN DEL IPML


,Variable,IPML,Nivel_IPML
0,Precipitacion_mm,1.0000,Muy Alta
9,ONI,1.0000,Muy Alta
6,Radiacion_Solar,1.0000,Muy Alta
4,Humedad_Relativa,0.9500,Muy Alta
24,Nivel_Minimo,0.9165,Muy Alta
34,pH,0.9095,Muy Alta
27,DEMANDA QUIMICA DE OXIGENO (DQO),0.9093,Muy Alta
2,Temp_Min_C,0.9000,Muy Alta
23,AGUAS RESIDUALES TRATADAS,0.8681,Alta
8,Mes,0.8668,Alta


In [34]:
#==========================================================================================
# RESUMEN DEL IPML
#==========================================================================================

print()

print("="*90)
print("RESUMEN DEL IPML")
print("="*90)

print()

print(universo["Nivel_IPML"].value_counts())


RESUMEN DEL IPML

Nivel_IPML
Baja        14
Muy Alta     8
Alta         8
Media        5
Name: count, dtype: int64


In [35]:
#==========================================================================================
# EXPORTACIÓN DEL RANKING IPML
#==========================================================================================

ranking_ipml = (

    universo

    .sort_values(

        "IPML",

        ascending=False

    )

    .reset_index(drop=True)

)

ranking_ipml.insert(

    0,

    "Ranking",

    ranking_ipml.index + 1

)

ranking_ipml.to_csv(

    "ranking_ipml_integracion_dimensiones.csv",

    index=False,

    encoding="utf-8-sig"

)

ranking_ipml.to_excel(

    "ranking_ipml_integracion_dimensiones.xlsx",

    index=False

)

print()

print("="*90)
print("ARCHIVOS GENERADOS")
print("="*90)

print()

print("ranking_ipml_integracion_dimensiones.csv")
print("ranking_ipml_integracion_dimensiones.xlsx")


ARCHIVOS GENERADOS

ranking_ipml_integracion_dimensiones.csv
ranking_ipml_integracion_dimensiones.xlsx


observación

Veo algo que conviene revisar antes de continuar

Por ejemplo:

* Nivel_Minimo tiene un IPML = 0.9165 aunque su cobertura es solo 16.48 %.
pH tiene un IPML = 0.9095 con una cobertura de 9.47 %.

* Esto ocurre porque la calidad pesa solo un 10 % en el índice. Metodológicamente no es un error, pero sí abre una decisión importante:

* Opción 1: Mantener el IPML como está y, en el proceso de selección, aplicar una regla de exclusión por calidad mínima (por ejemplo, cobertura ≥ 30 % o ≥ 50 %). Así el índice refleja la importancia de la variable y la decisión final incorpora un criterio de viabilidad de los datos.

* Opción 2: aumentar el peso de la calidad dentro del IPML, lo que penalizaría mucho más a las variables con baja cobertura.

Considero más sólido el primer enfoque: separar el ranking de prioridad de las reglas de selección final. De esa manera el IPML mide el potencial de una variable y el proceso de selección aplica criterios operativos para construir el dataset definitivo.

## **M4. Selección del Universo Definitivo para Machine Learning**

In [ ]:
#==========================================================================================
# SCRIPT
# SELECCIÓN DEL UNIVERSO DEFINITIVO PARA MACHINE LEARNING
#==========================================================================================
#
# Objetivo:
#
# Seleccionar las variables que conformarán el Dataset definitivo para
# Machine Learning utilizando el Índice de Priorización (IPML) y criterios
# adicionales de calidad y cobertura.
#
# Producto esperado:
#
# • Dataset de variables seleccionadas
# • Ranking final de variables
#
#==========================================================================================

In [36]:
import os
import pandas as pd

# Definir las rutas
url = "https://raw.githubusercontent.com/jriatiga/FRAMEWORK_V7/refs/heads/main/DATA/MASTER/C11_MASTER/Universo_Analitico/ranking_ipml_integracion_dimensiones.csv"
ruta_local_ranking = "ranking_ipml_integracion_dimensiones.csv"  # Ajusta la ruta si lo tienes en otra carpeta

try:
    # Intentar cargar desde GitHub
    print("Intentando cargar el ranking IPML desde GitHub...")
    ranking = pd.read_csv(url)
    print("¡Carga exitosa desde GitHub!")
except Exception as e:
    print(f"No se pudo cargar desde GitHub ({e}). Intentando cargar desde el archivo local...")
    # Cargar desde la ruta local si falla la web
    ranking = pd.read_csv(ruta_local_ranking)
    print("¡Carga exitosa desde el archivo local!")

print()
print("=" * 90)
print("RANKING IPML")
print("=" * 90)
print()

print(f"Variables : {len(ranking)}")

display(ranking.head())

Intentando cargar el ranking IPML desde GitHub...
No se pudo cargar desde GitHub (HTTP Error 404: Not Found). Intentando cargar desde el archivo local...
¡Carga exitosa desde el archivo local!

RANKING IPML

Variables : 35


,Ranking,Variable,Tipo,Relaciones_Fuertes,Indice_Redundancia,Nivel_Redundancia,Diagnostico_Redundancia,VIF,Estado_VIF,Diagnostico_VIF,Recomendacion_VIF,Categoria,Importancia,Peso,Puntaje_Conocimiento_Dominio,% Cobertura,Puntaje_Calidad,Puntaje_Redundancia,Puntaje_Multicolinealidad,IPML,Nivel_IPML
0,1,Precipitacion_mm,Numérica,0,0.0,Muy Baja,Variable prácticamente independiente.,1.898,Baja,Variable con baja dependencia estructural.,Variable estable para Machine Learning.,Climática,Muy Alta,3,1.0,100.00,1.000,1.0,1.00,1.0000,Muy Alta
1,2,ONI,Numérica,0,0.0,Muy Baja,Variable prácticamente independiente.,1.265,Baja,Variable con baja dependencia estructural.,Variable estable para Machine Learning.,Variabilidad Climática,Muy Alta,3,1.0,100.00,1.000,1.0,1.00,1.0000,Muy Alta
2,3,Radiacion_Solar,Numérica,0,0.0,Muy Baja,Variable prácticamente independiente.,1.704,Baja,Variable con baja dependencia estructural.,Variable estable para Machine Learning.,Climática,Muy Alta,3,1.0,100.00,1.000,1.0,1.00,1.0000,Muy Alta
3,4,Humedad_Relativa,Numérica,0,0.0,Muy Baja,Variable prácticamente independiente.,7.128,Moderada,Existe una dependencia moderada aceptable para...,Sin restricciones importantes.,Climática,Muy Alta,3,1.0,100.00,1.000,1.0,0.75,0.9500,Muy Alta
4,5,Nivel_Minimo,Numérica,0,0.0,Muy Baja,Variable prácticamente independiente.,1.027,Baja,Variable con baja dependencia estructural.,Variable estable para Machine Learning.,Hidrológica,Muy Alta,3,1.0,16.48,0.165,1.0,1.00,0.9165,Muy Alta


In [37]:
#==========================================================================================
# CRITERIOS DE SELECCIÓN
#==========================================================================================

COBERTURA_MINIMA = 30

IPML_MINIMO = 0.80

print()

print("="*90)
print("CRITERIOS DEL FRAMEWORK")
print("="*90)

print()

print(f"IPML mínimo              : {IPML_MINIMO}")

print(f"Cobertura mínima (%)     : {COBERTURA_MINIMA}")

print("Estado VIF permitido     : diferente de 'Perfecta'")


CRITERIOS DEL FRAMEWORK

IPML mínimo              : 0.8
Cobertura mínima (%)     : 30
Estado VIF permitido     : diferente de 'Perfecta'


In [38]:
#==========================================================================================
# ESTADO FINAL DE LAS VARIABLES
#==========================================================================================

def clasificar_variable(fila):

    #----------------------------------------------------------
    # Validar IPML no calculado
    #----------------------------------------------------------

    if pd.isna(fila["IPML"]):
        return "No Evaluada"

    #----------------------------------------------------------

    if fila["IPML"] < IPML_MINIMO:
        return "No Prioritaria"

    if fila["% Cobertura"] < COBERTURA_MINIMA:
        return "Pendiente por Cobertura"

    if fila["Estado_VIF"] == "Perfecta":
        return "Revisar por Multicolinealidad"

    return "Seleccionada"


ranking["Estado_Final"] = ranking.apply(

    clasificar_variable,

    axis=1

)

display(

    ranking[
        [
            "Variable",
            "IPML",
            "% Cobertura",
            "Estado_VIF",
            "Estado_Final"
        ]
    ].head(20)

)

,Variable,IPML,% Cobertura,Estado_VIF,Estado_Final
0,Precipitacion_mm,1.0000,100.00,Baja,Seleccionada
1,ONI,1.0000,100.00,Baja,Seleccionada
2,Radiacion_Solar,1.0000,100.00,Baja,Seleccionada
3,Humedad_Relativa,0.9500,100.00,Moderada,Seleccionada
4,Nivel_Minimo,0.9165,16.48,Baja,Pendiente por Cobertura
5,pH,0.9095,9.47,Baja,Pendiente por Cobertura
6,DEMANDA QUIMICA DE OXIGENO (DQO),0.9093,9.28,Baja,Pendiente por Cobertura
7,Temp_Min_C,0.9000,100.00,Alta,Seleccionada
8,AGUAS RESIDUALES TRATADAS,0.8681,18.18,Baja,Pendiente por Cobertura
9,Mes,0.8668,100.00,Baja,Seleccionada


In [39]:
#==========================================================================================
# RESUMEN DE LA SELECCIÓN
#==========================================================================================

print()

print("="*90)
print("RESUMEN DE LA SELECCIÓN")
print("="*90)

print()

print(

    ranking["Estado_Final"]

    .value_counts()

)


RESUMEN DE LA SELECCIÓN

Estado_Final
No Prioritaria             18
Seleccionada                9
Pendiente por Cobertura     7
No Evaluada                 1
Name: count, dtype: int64


In [40]:
#==========================================================================================
# VARIABLES SELECCIONADAS
#==========================================================================================

variables_ml = ranking[

    ranking["Estado_Final"] == "Seleccionada"

].copy()

print()

print("="*90)
print("VARIABLES PARA MACHINE LEARNING")
print("="*90)

print()

print(f"Variables seleccionadas : {len(variables_ml)}")

display(

    variables_ml[
        [
            "Ranking",
            "Variable",
            "IPML",
            "Estado_Final"
        ]
    ]

)


VARIABLES PARA MACHINE LEARNING

Variables seleccionadas : 9


,Ranking,Variable,IPML,Estado_Final
0,1,Precipitacion_mm,1.0000,Seleccionada
1,2,ONI,1.0000,Seleccionada
2,3,Radiacion_Solar,1.0000,Seleccionada
3,4,Humedad_Relativa,0.9500,Seleccionada
7,8,Temp_Min_C,0.9000,Seleccionada
9,10,Mes,0.8668,Seleccionada
10,11,Velocidad_Viento,0.8668,Seleccionada
11,12,irca,0.8499,Seleccionada
12,13,VolumenUtilDiarioMasa,0.8499,Seleccionada



Desde el punto de vista estadístico, el resultado es correcto.

Pero desde el punto de vista de gestión hídrica, me preocupa una cosa.

Prácticamente todas las variables seleccionadas pertenecen a:

* Climáticas
* Temporales
* Hidrológicas

y solamente quedó una variable importante de calidad del agua:

* irca

Mientras que quedaron por fuera variables como:

* pH
* DQO
* Oxígeno Disuelto
* Conductividad
* DBO5

No porque sean malas variables, sino por la cobertura.

In [41]:
#==========================================================================================
# DISTRIBUCIÓN DE VARIABLES SELECCIONADAS POR CATEGORÍA
#==========================================================================================

print()

print("="*90)
print("VARIABLES SELECCIONADAS POR CATEGORÍA")
print("="*90)

display(

    variables_ml[
        [
            "Categoria",
            "Variable"
        ]

    ].groupby(

        "Categoria"

    ).count()

)


VARIABLES SELECCIONADAS POR CATEGORÍA


,Variable
Categoria,
Calidad del Agua,1
Climática,5
Hidrológica,1
Temporal,1
Variabilidad Climática,1


In [42]:
display(

    variables_ml.groupby(

        "Importancia"

    )["Variable"]

    .count()

)

,Variable
Importancia,
Alta,2
Muy Alta,7


In [43]:
#==========================================================================================
# VARIABLES SELECCIONADAS
#==========================================================================================

variables_ml = ranking[
    ranking["Estado_Final"] == "Seleccionada"
].copy()

print()

print("="*90)
print("VARIABLES PARA MACHINE LEARNING")
print("="*90)

print()

print(f"Variables seleccionadas : {len(variables_ml)}")

display(

    variables_ml[
        [
            "Ranking",
            "Variable",
            "IPML",
            "Estado_Final"
        ]
    ]

)


VARIABLES PARA MACHINE LEARNING

Variables seleccionadas : 9


,Ranking,Variable,IPML,Estado_Final
0,1,Precipitacion_mm,1.0000,Seleccionada
1,2,ONI,1.0000,Seleccionada
2,3,Radiacion_Solar,1.0000,Seleccionada
3,4,Humedad_Relativa,0.9500,Seleccionada
7,8,Temp_Min_C,0.9000,Seleccionada
9,10,Mes,0.8668,Seleccionada
10,11,Velocidad_Viento,0.8668,Seleccionada
11,12,irca,0.8499,Seleccionada
12,13,VolumenUtilDiarioMasa,0.8499,Seleccionada


In [44]:
#==========================================================================================
# VARIABLES PENDIENTES
#==========================================================================================

variables_pendientes = ranking[
    ranking["Estado_Final"] == "Pendiente por Cobertura"
].copy()

print()

print("="*90)
print("VARIABLES PENDIENTES")
print("="*90)

print()

print(f"Variables pendientes : {len(variables_pendientes)}")

display(

    variables_pendientes[
        [
            "Ranking",
            "Variable",
            "IPML",
            "% Cobertura",
            "Estado_Final"
        ]
    ]

)


VARIABLES PENDIENTES

Variables pendientes : 7


,Ranking,Variable,IPML,% Cobertura,Estado_Final
4,5,Nivel_Minimo,0.9165,16.48,Pendiente por Cobertura
5,6,pH,0.9095,9.47,Pendiente por Cobertura
6,7,DEMANDA QUIMICA DE OXIGENO (DQO),0.9093,9.28,Pendiente por Cobertura
8,9,AGUAS RESIDUALES TRATADAS,0.8681,18.18,Pendiente por Cobertura
13,14,CALIDAD DEL AGUA,0.8183,18.18,Pendiente por Cobertura
14,15,OXIGENO DISUELTO (OD),0.8094,9.47,Pendiente por Cobertura
15,16,DEMANDA BIOQUIMICA DE OXIGENO (DBO5),0.8050,4.92,Pendiente por Cobertura


In [45]:
#==========================================================================================
# VARIABLES EXCLUIDAS
#==========================================================================================

variables_excluidas = ranking[
    ranking["Estado_Final"].isin(
        [
            "No Prioritaria",
            "Revisar por Multicolinealidad",
            "Excluida por Varianza Nula"
        ]
    )
].copy()

print()

print("="*90)
print("VARIABLES EXCLUIDAS")
print("="*90)

print()

print(f"Variables excluidas : {len(variables_excluidas)}")

display(

    variables_excluidas[
        [
            "Ranking",
            "Variable",
            "IPML",
            "Estado_Final"
        ]
    ]

)


VARIABLES EXCLUIDAS

Variables excluidas : 18


,Ranking,Variable,IPML,Estado_Final
16,17,Temp_Max_C,0.7999,No Prioritaria
17,18,Temp_Media_C,0.7999,No Prioritaria
18,19,TEMPERATURA,0.7262,No Prioritaria
19,20,SOLIDOS SUSPENDIDOS TOTALES,0.7262,No Prioritaria
20,21,TURBIDEZ,0.7231,No Prioritaria
21,22,Anio,0.6832,No Prioritaria
22,23,CONTINUIDAD DE ACUEDUCTO URBANO,0.6350,No Prioritaria
23,24,FOSFORO TOTAL,0.6260,No Prioritaria
24,25,NITROGENO TOTAL,0.6222,No Prioritaria
25,26,CONDUCTIVIDAD ELECTRICA,0.5264,No Prioritaria


In [46]:
#==========================================================================================
# JUSTIFICACIÓN METODOLÓGICA
#==========================================================================================

def justificar_estado(estado):

    justificaciones = {

        "Seleccionada":
        "Cumple todos los criterios del Framework.",

        "Pendiente por Cobertura":
        "Cobertura inferior al umbral mínimo establecido.",

        "Revisar por Multicolinealidad":
        "Presenta multicolinealidad perfecta.",

        "No Prioritaria":
        "IPML inferior al umbral definido.",

        "Excluida por Varianza Nula":
        "Variable constante que no aporta información al modelo."

    }

    return justificaciones.get(estado, "")


ranking["Justificacion"] = ranking["Estado_Final"].apply(justificar_estado)

In [47]:
#==========================================================================================
# EXPORTACIÓN DE RESULTADOS
#==========================================================================================

import os
from pathlib import Path

ruta_str = "Preparacion_Machine_Learning/ipml/"

# Crear el directorio si no existe
ruta = Path(ruta_str)
ruta.mkdir(parents=True, exist_ok=True)

ranking.to_csv(
    ruta / "clasificacion_ipml.csv",
    index=False,
    encoding="utf-8-sig"
)

ranking.to_excel(
    ruta / "clasificacion_ipml.xlsx",
    index=False
)

variables_ml.to_csv(
    ruta / "variables_machine_learning.csv",
    index=False,
    encoding="utf-8-sig"
)

variables_ml.to_excel(
    ruta / "variables_machine_learning.xlsx",
    index=False
)

variables_pendientes.to_csv(
    ruta / "variables_pendientes.csv",
    index=False,
    encoding="utf-8-sig"
)

variables_pendientes.to_excel(
    ruta / "variables_pendientes.xlsx",
    index=False
)

variables_excluidas.to_csv(
    ruta / "variables_excluidas.csv",
    index=False,
    encoding="utf-8-sig"
)

variables_excluidas.to_excel(
    ruta / "variables_excluidas.xlsx",
    index=False
)

print()

print("="*90)
print("ARCHIVOS GENERADOS")
print("="*90)

print()

print("clasificacion_ipml.csv")
print("clasificacion_ipml.xlsx")
print("variables_machine_learning.csv")
print("variables_machine_learning.xlsx")
print("variables_pendientes.csv")
print("variables_pendientes.xlsx")
print("variables_excluidas.csv")
print("variables_excluidas.xlsx")


ARCHIVOS GENERADOS

clasificacion_ipml.csv
clasificacion_ipml.xlsx
variables_machine_learning.csv
variables_machine_learning.xlsx
variables_pendientes.csv
variables_pendientes.xlsx
variables_excluidas.csv
variables_excluidas.xlsx
